diagnostics-app/src/app/core/services/routine-v2.service.ts

```diff
@@ -0,0 +1,237 @@
+ * capturing events from Chrome extension.
+ */
+
+import {Injectable} from '@angular/core';
+import {Subject} from 'rxjs';
+
+import {VISIBLE_ROUTINE_V2_CARDS} from 'common/config';
+import {
+  CancelRoutineMessage,
+  CreateRoutineMessage,
+  IsCreateRoutineResponse,
+  IsRoutineArgumentSupportedMessage,
+  IsRoutineSupportStatusInfo,
+  PortName,
+  Request,
+  RequestType,
+  Response,
+  ResponseErrorInfoMessage,
+  RoutineV2Action,
+  RoutineV2Argument,
+  RoutineV2Event,
+  RoutineV2Request,
+  RoutineV2Response,
+  StartRoutineMessage,
+} from 'common/message';
+import {
+  CreateRoutineResponse,
+  RoutineSupportStatus,
+  RoutineSupportStatusInfo,
+} from 'common/telemetry-extension-types';
+import {environment} from 'environments/environment';
+
+export interface getSubjectResponse {
+  success: Boolean;
+  subject?: Subject<RoutineV2Event>;
+  error?: string;
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export class RoutineV2Service {
+  // the ID of the extension service connects to.
+  private extensionId: string = environment.extensionId;
+  // A map storing subjects to forward the runtime responses from running routine V2 to.
+  private subjects = new Map<RoutineV2Argument, Subject<RoutineV2Event>>();
+  // the port for connecting with the extension to receive runtime responses.
+  private port?: chrome.runtime.Port;
+  // Since we use routine argument as the source of truth for each routine type,
+  // we store a map which maps between the uuid and its corresponding routine
+  // argument.
+  private uuidToRoutineArgument = new Map<string, RoutineV2Argument>();
+  // A cache to preload the supported routines.
+  private supportabilityCache = new Map<
+    RoutineV2Argument,
+    Promise<RoutineSupportStatusInfo>
+  >();
+
+  private constructRoutineV2Request: (payload: RoutineV2Request) => Request = (
+    payload,
+  ) => {
+    return {type: RequestType.ROUTINE_V2, routineV2: payload};
+  };
+
+  private sendRequest: (request: Request) => Promise<RoutineV2Response> = (
+    request: Request,
+  ) => {
+    return new Promise((resolve, reject) => {
+      try {
+        window.chrome.runtime.sendMessage(
+          this.extensionId,
+          request,
+          (response: Response) => {
+            if (!response.success) {
+              reject(response.error);
+            } else if (!response.routineV2) {
+              throw 'Invalid response';
+            } else {
+              resolve(response.routineV2);
+            }
+          },
+        );
+      } catch (err) {
+        reject(err);
+      }
+    });
+  };
+
+  Init(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      try {
+        this.port = window.chrome.runtime.connect(this.extensionId, {
+          name: PortName.ROUTINE_V2_PORT,
+        });
+        this.port.onMessage.addListener((msg: RoutineV2Event) => {
+          if (
+            'uuid' in msg.event &&
+            this.uuidToRoutineArgument.has(msg.event.uuid!)
```

Comment:
Does `if(this.uuidToRoutineArgument.has(msg.event.uuid?))` work?

Reply:
I don't think so. The JS map has string as key type, and cannot be queried using `undefined` type as key.

We must use `msg.event.uuid!` to tell compiler the value exists. And we should only do that if we first verify that `uuid` is a valid key.